bin/build_cmd.ml

```diff
@@ -1,21 +1,64 @@
 open Stdune
 open Import
 
-let run_build_command ~common ~targets =
+let make_setup common =
+  let rpc = Common.rpc common in
+  let build_mutex = Option.map rpc ~f:Dune_rules.Dune_rpc_server.build_mutex in
+  Import.Main.setup ?build_mutex common
+
+let run_build_command_poll ~(common : Common.t) ~targets ~setup =
+  let open Fiber.O in
   let once () =
-    Memo.Build.run
-      (let open Memo.Build.O in
-      let* setup = Import.Main.setup common in
-      do_build (targets setup))
+    Cached_digest.invalidate_cached_timestamps ();
+    let* setup = Memo.Build.run (setup ()) in
+    match
+      match
+        let open Option.O in
+        let* rpc = Common.rpc common in
+        Dune_rules.Dune_rpc_server.pending_build_action rpc
+      with
+      | None -> `Build (targets setup, None)
+      | Some Shutdown -> `Shutdown
+      | Some (Build (targets, ivar)) ->
+        `Build (Target.resolve_targets_exn common setup targets, Some ivar)
+    with
+    | `Shutdown -> Fiber.return `Stop
+    | `Build (targets, ivar) ->
+      let* () =
+        match ivar with
+        | None -> Fiber.return ()
+        | Some ivar -> Fiber.Ivar.fill ivar Accepted
```

Comment:
IIRC, it works as follow: `dune build -w x` will initially build `x`. Then if  a request to build `y` arrives, dune will temporarily switch to building `y`. Once it has finished building `y`, it will go back to building `x`.

Intuitively, I would expect the following:

- initially the goal is `x`
- then when the RPC request to build `y` arrives, the goal becomes `x y`. This might happen in the middle of a build, in which case we immediately start building `y` as well without waiting for the end of the current build
- once the build of `y` has succeeded at least once, the goal goes back to just `x`

We should be able to do that with a memo cell and the newly introduced invalidation, for instance:

```ocaml
module Goal : sig
   val set : unit Action_builder.t -> unit

   val get : unit -> unit Action_builder.t Memo.Build.t
end = struct
  let current = ref (Action_builder.return ())

  let memo =
    Memo.create "goal" ~doc:"current build goal"
      ~input:(module Unit)
      ~output:(Simple (module Action_builder))
      ~visibility:Hidden Async
      (fun () -> Memo.Build.return !current)

  let get = Memo.exec memo 

  let set x = current := x; Memo.Cell.invalidate (Memo.cell memo ())
end
```

Reply:
> Intuitively, I would expect the following:

To me, the mechanism you're proposing works better when an RPC call is made to build something in the background. For example, I can imagine rotor or merlin requesting some cmi's to be built this way. It works less well for the use case I've had in mind - to avoid having to restart `$ dune build -w` as part of the workflow. I would still need to restart things if I want to change building this to `$ dune build @lib/all` for example.

Reply:
From what you are describing, I understand is that you want to change the goal entirely to a new one. Which is not what the PR is currently implementing. From what I understand, what is implemented is temporarily changing the goal to an alternative one and then going back to the old one.